spring-cloud-dataflow-admin/src/main/java/org/springframework/cloud/dataflow/admin/controller/StreamController.java

```diff
@@ -222,34 +222,77 @@ private void deployStream(StreamDefinition stream, Map<String, String> cumulated
-			isDownStreamModulePartitioned = isPartitionedConsumer(currentModule, moduleDeploymentProperties,
-					upstreamModuleSupportsPartition);
+		}
+	}
+
+	/**
+	 * Maybe update the {@literal currentModule} {@literal moduleDeploymentProperties} with properties related
+	 * to partitioning.
+	 *
+	 * @param stream the whole stream definition
+	 * @param cumulatedDeploymentProperties the whole set of stream deployment properties
+	 * @param info information about the previous (i.e. downstream) module in the stream
+	 * @param currentModule the module whose deployment properties we want to alter
+	 * @param moduleDeploymentProperties the deployment properties for the current module
+	 * @return information about the current module, to be used as 'previous' in the next invocation
+	 */
+	private ModulePartitionInfo postProcessPartitioningProperties(StreamDefinition stream, Map<String, String> cumulatedDeploymentProperties, ModulePartitionInfo info, ModuleDefinition currentModule, Map<String, String> moduleDeploymentProperties) {
+		boolean upstreamModuleSupportsPartition = upstreamModuleHasPartitionInfo(stream, currentModule, cumulatedDeploymentProperties);
+		// consumer module partition properties
+		if (isPartitionedConsumer(currentModule, moduleDeploymentProperties, upstreamModuleSupportsPartition)) {
+			updateConsumerPartitionProperties(moduleDeploymentProperties);
+		}
+		// producer module partition properties
+		if (info.isPartitioned) {
+			updateProducerPartitionProperties(moduleDeploymentProperties, info.instances);
+		}
+		int nextModuleCount = getNextModuleCount(moduleDeploymentProperties);
+		boolean isDownStreamModulePartitioned = isPartitionedConsumer(currentModule, moduleDeploymentProperties,
+				upstreamModuleSupportsPartition);
+		return new ModulePartitionInfo(nextModuleCount, isDownStreamModulePartitioned);
+	}
+
+	private ModuleCoordinates getModuleCoordinates(StreamDefinition.DeploymentOrderIterator iterator, ModuleDefinition currentModule) {
+		ModuleType type = determineModuleType(currentModule, !iterator.hasMoreUpstream(), !iterator.hasMoreDownstream());
+		ModuleRegistration registration = this.registry.find(currentModule.getName(), type);
+		if (registration == null) {
+			throw new IllegalArgumentException(String.format(
+					"Module %s of type %s not found in registry", currentModule.getName(), type));
+		}
+		return registration.getCoordinates();
+	}
+
+	/**
+	 * Holds partitioning state about a module in the context of deployment.
+	 */
+	private static class ModulePartitionInfo {
+		private final int instances;
+		private final boolean isPartitioned;
+
+		private ModulePartitionInfo(int instances, boolean isPartitioned) {
+			this.instances = instances;
+			this.isPartitioned = isPartitioned;
+		}
+
+		/**
+		 * Return an info suitable for the initial module, <i>i.e.</i> the first one to be deployed.
+		 */
+		private static ModulePartitionInfo initial() {
```

Comment:
I think "initial" also violates the clarity we're trying to achieve with regard to order (sounds like the FIRST module in a stream to me... as opposed to first to be deployed).

Also, we need to consider how we'll handle a situation where a stream definition ends with a named channel and yet partitioning is expected.